pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/rebalance/DefaultRebalanceSegmentStrategy.java

```diff
@@ -0,0 +1,300 @@
+      try {
+        targetNumReplicas = Integer.parseInt(replicasString);
+      } catch (Exception e) {
+        throw new RuntimeException("Invalid value for replicasPerPartition:'" + replicasString + "'", e);
+      }
+    } else {
+      targetNumReplicas = Integer.parseInt(tableConfig.getValidationConfig().getReplication());
+    }
+    rebalanceServing(idealState, tableConfig, targetNumReplicas);
+    boolean dryRun =
+        Boolean.valueOf(rebalanceUserParams.getConfig(RebalanceUserParamConstants.DRYRUN, DEFAULT_DRY_RUN));
+    if (!dryRun) {
+      LOGGER.info("Updating ideal state for table {}", tableNameWithType);
+      updateIdealState(tableNameWithType, targetNumReplicas,
+          idealState.getRecord().getMapFields());
+    } else {
+      LOGGER.info("Dry run. Skip writing ideal state");
+    }
+    return idealState;
+  }
+
+
+
+  /**
+   * Rebalances serving segments based on autorebalance strategy
+   * @param idealState
+   * @param tableConfig
+   * @param targetNumReplicas
+   * @return
+   */
+  protected IdealState rebalanceServing(IdealState idealState, TableConfig tableConfig, int targetNumReplicas) {
+
+    LOGGER.info("Rebalancing serving segments for table {}", tableConfig.getTableName());
+    TagConfig tagConfig = new TagConfig(tableConfig, _helixManager);
+    List<String> servingInstances =
+        _helixAdmin.getInstancesInClusterWithTag(_helixClusterName, tagConfig.getCompletedServerTag());
+    List<String> enabledServingInstances =
+        HelixHelper.getEnabledInstancesWithTag(_helixAdmin, _helixClusterName, tagConfig.getCompletedServerTag());
+
+    int numReplicasInIdealState = Integer.parseInt(idealState.getReplicas());
+
+    if (numReplicasInIdealState > targetNumReplicas) { // We need to reduce the number of replicas per helix partition.
+
+      for (String segmentId : idealState.getPartitionSet()) {
+        Map<String, String> instanceStateMap = idealState.getInstanceStateMap(segmentId);
+        if (instanceStateMap.size() > targetNumReplicas) {
+          Set<String> keys = instanceStateMap.keySet();
+          while (instanceStateMap.size() > targetNumReplicas) {
+            instanceStateMap.remove(keys.iterator().next());
+          }
+        } else if (instanceStateMap.size() < targetNumReplicas) {
+          LOGGER.warn("Table {}, segment {} has {} replicas, less than {} (requested number of replicas",
+              idealState.getResourceName(), segmentId, instanceStateMap.size(), targetNumReplicas);
+        }
+      }
+    } else { // Number of replicas is either the same or higher, so invoke Helix rebalancer.
+
+      final Map<String, Map<String, String>> mapFields = idealState.getRecord().getMapFields();
+
+      List<Map.Entry<String, Map<String, String>>> removedEntries = new LinkedList<>();
+      if (tableConfig.getTableType().equals(CommonConstants.Helix.TableType.REALTIME)) {
+        // FIXME: what about those in OFFLINE states?
```

Comment:
Once you get the names of the latest segments for each partition, you can mark them as CONSUMING state even if they are OFFLINE in the current IDEALSTATE record. The worst that will happen is that they go back to OFFLINE state because of some issues with consumption